### PR TITLE
Disable automatic logs injection during IIS PreStartInit phase

### PR DESCRIPTION
**Pull request by zacharycmontoya**
Replaces #1139

### Issue
This change fixes a crash that occurs under the following conditions:
1. .NET Framework application is hosted in IIS
2. Automatic logs injection is enabled
3. Application uses the log4net logging library
4. Automatic instrumentation is triggered during the IIS PreStartInit phase

### Context
In order to store ambient information for the logging context on .NET Framework, log4net uses `CallContext`. The first time a property is stored, a log4net dictionary type is created and permanently stored in the `CallContext`, even if the dictionary is later found to be empty. This creates an error condition: If this is stored into the `CallContext` during the IIS PreStartInit phase (which occurs in the application-specific AppDomain), then a crash will occur when the phase is finished and execution resumes in the IIS default AppDomain.

### Solution
The solution is two-fold:

1. When automatic instrumentation is enabled, we already perform IL rewriting at the beginning of the IIS PreStartInit code path to load our managed code. Modify this code path to begin and end with a call to `AppDomain.SetData` to set a flag that indicates if the IIS PreStartInit phase is still running. This flag can be observed from elsewhere in the AppDomain and need not be on the same thread.
2. Register to a new `TraceStarted` event provided by the ScopeManager and register a callback to evaluate whether we're still running in this PreStartInit phase. It will either check for the AppDomain flag (if it was not-null) or it will fallback to checking the StackTrace (which is expensive). As soon as we detect false, we won't execute these checks again.

Optimizations to this strategy include never subscribing to the `TraceStarted` event if one of the following conditions is met at Tracer startup:
1. Automatic instrumentation is enabled and the IIS PreStartInit flag is already false
2. No automatic instrumentation is enabled and the process name is not `w3wp` or `iisexpress`

### Testing

1. Added automatic instrumentation + logs injection to IIS smoke test in CI/CD (project: `Samples.AspNet472.LoaderOptimizationRegKey`)
2. Tested locally with logs injection enabled and traces started in the IIS PreInitStart phase for both 1) IIS | IIS Express and 2) Automatic Instrumentation | Custom Instrumentation

### Benchmarks
Observations from the below numbers:
- The only performance penalty is when LogsInjectionEnabled=true and starting multiple traces inside the IIS PreStartInit phase of execution. Though, we expect this to be the rare case anyways.
- `AppDomain.SetData` / `AppDomain.GetData` check, which is used when automatic instrumentation is enabled, is at least 2x faster than the StackTrace walk. The StackTrace check is only needed when automatic instrumentation is not enabled, so this should be the rare case.

Code without the Tracer and scenario setup done in `[GlobalSetup]` before the iteration:
```
        [Benchmark]
        public int OpenTracesWithLogsInjectionEnabled()
        {
            int retVal = 0;

            for (int i = 0; i < NumberOfTraces; i++)
            {
                using (var scope = tracer.StartActive("iteration"))
                {
                    retVal++;
                }
            }

            return retVal;
        }
```

#### Scenario: LogsInjectionEnabled=false

|              State |     AutomaticInstr |     NumberOfTraces |           Mean |          Error |         StdDev |         Median |       Gen 0 |     Allocated |
|------------------- |------------------- |------------------- |---------------:|---------------:|---------------:|---------------:|------------:|--------------:|
|         **NotIIS** |          **False** |              **5** |   **4.203 μs** |  **0.0837 μs** |  **0.1941 μs** |   **4.209 μs** |  **0.6180** |   **2.86 KB** |
|         **NotIIS** |          **False** |             **10** |   **8.185 μs** |  **0.1601 μs** |  **0.2138 μs** |   **8.057 μs** |  **1.2360** |   **5.72 KB** |
|         **NotIIS** |          **False** |             **25** |  **20.413 μs** |  **0.2383 μs** |  **0.2113 μs** |  **20.355 μs** |  **3.0823** |   **14.3 KB** |
|         **NotIIS** |           **True** |              **5** |   **4.196 μs** |  **0.0834 μs** |  **0.1084 μs** |   **4.160 μs** |  **0.6180** |   **2.86 KB** |
|         **NotIIS** |           **True** |             **10** |   **8.190 μs** |  **0.0611 μs** |  **0.0477 μs** |   **8.194 μs** |  **1.2360** |   **5.72 KB** |
|         **NotIIS** |           **True** |             **25** |  **20.724 μs** |  **0.2628 μs** |  **0.2458 μs** |  **20.731 μs** |  **3.0823** |   **14.3 KB** |
|                    |                    |                    |                |                |                |                |             |               |
| **IISNotPreStart** |          **False** |              **5** |   **4.099 μs** |  **0.0754 μs** |  **0.0706 μs** |   **4.070 μs** |  **0.6180** |   **2.86 KB** |
| **IISNotPreStart** |          **False** |             **10** |   **8.343 μs** |  **0.0832 μs** |  **0.0778 μs** |   **8.351 μs** |  **1.2360** |   **5.72 KB** |
| **IISNotPreStart** |          **False** |             **25** |  **20.504 μs** |  **0.4028 μs** |  **0.4795 μs** |  **20.407 μs** |  **3.0823** |   **14.3 KB** |
| **IISNotPreStart** |           **True** |              **5** |   **4.198 μs** |  **0.0789 μs** |  **0.0844 μs** |   **4.232 μs** |  **0.6180** |   **2.86 KB** |
| **IISNotPreStart** |           **True** |             **10** |   **8.106 μs** |  **0.1070 μs** |  **0.0894 μs** |   **8.075 μs** |  **1.2360** |   **5.72 KB** |
| **IISNotPreStart** |           **True** |             **25** |  **20.479 μs** |  **0.4094 μs** |  **0.7383 μs** |  **20.158 μs** |  **3.0823** |   **14.3 KB** |
|                    |                    |                    |                |                |                |                |             |               |
|    **IISPreStart** |          **False** |              **5** |   **4.228 μs** |  **0.0805 μs** |  **0.1569 μs** |   **4.193 μs** |  **0.6180** |   **2.86 KB** |
|    **IISPreStart** |          **False** |             **10** |   **8.399 μs** |  **0.1649 μs** |  **0.3586 μs** |   **8.288 μs** |  **1.2360** |   **5.72 KB** |
|    **IISPreStart** |          **False** |             **25** |  **20.874 μs** |  **0.3948 μs** |  **0.4388 μs** |  **20.754 μs** |  **3.0823** |   **14.3 KB** |
|    **IISPreStart** |           **True** |              **5** |   **4.134 μs** |  **0.0815 μs** |  **0.1245 μs** |   **4.115 μs** |  **0.6180** |   **2.86 KB** |
|    **IISPreStart** |           **True** |             **10** |   **8.097 μs** |  **0.0651 μs** |  **0.0543 μs** |   **8.096 μs** |  **1.2360** |   **5.72 KB** |
|    **IISPreStart** |           **True** |             **25** |  **21.767 μs** |  **0.3984 μs** |  **0.3727 μs** |  **21.721 μs** |  **3.0823** |   **14.3 KB** |

#### Scenario: LogsInjectionEnabled=true

|              State |     AutomaticInstr |     NumberOfTraces |           Mean |          Error |         StdDev |         Median |       Gen 0 |     Allocated |
|------------------- |------------------- |------------------- |---------------:|---------------:|---------------:|---------------:|------------:|--------------:|
|         **NotIIS** |          **False** |              **5** |   **7.768 μs** |  **0.0961 μs** |  **0.0852 μs** |   **7.772 μs** |  **1.4343** |   **6.66 KB** |
|         **NotIIS** |          **False** |             **10** |  **15.735 μs** |  **0.2832 μs** |  **0.2365 μs** |  **15.757 μs** |  **2.8687** |  **13.32 KB** |
|         **NotIIS** |          **False** |             **25** |  **40.418 μs** |  **0.7801 μs** |  **1.4460 μs** |  **40.113 μs** |  **7.2021** |   **33.3 KB** |
|         **NotIIS** |           **True** |              **5** |   **8.021 μs** |  **0.1447 μs** |  **0.1283 μs** |   **8.003 μs** |  **1.4343** |   **6.66 KB** |
|         **NotIIS** |           **True** |             **10** |  **15.484 μs** |  **0.2796 μs** |  **0.2615 μs** |  **15.387 μs** |  **2.8687** |  **13.32 KB** |
|         **NotIIS** |           **True** |             **25** |  **39.060 μs** |  **0.7682 μs** |  **0.8220 μs** |  **38.811 μs** |  **7.2021** |   **33.3 KB** |
|                    |                    |                    |                |                |                |                |             |               |
| **IISNotPreStart** |          **False** |              **5** |   **7.911 μs** |  **0.1498 μs** |  **0.1784 μs** |   **7.895 μs** |  **1.4343** |   **6.66 KB** |
| **IISNotPreStart** |          **False** |             **10** |  **15.590 μs** |  **0.3103 μs** |  **0.3573 μs** |  **15.626 μs** |  **2.8687** |  **13.32 KB** |
| **IISNotPreStart** |          **False** |             **25** |  **38.816 μs** |  **0.5025 μs** |  **0.4454 μs** |  **38.716 μs** |  **7.2021** |   **33.3 KB** |
| **IISNotPreStart** |           **True** |              **5** |   **7.775 μs** |  **0.0994 μs** |  **0.0881 μs** |   **7.756 μs** |  **1.4343** |   **6.66 KB** |
| **IISNotPreStart** |           **True** |             **10** |  **16.035 μs** |  **0.3162 μs** |  **0.4634 μs** |  **15.901 μs** |  **2.8687** |  **13.32 KB** |
| **IISNotPreStart** |           **True** |             **25** |  **40.042 μs** |  **0.5196 μs** |  **0.4339 μs** |  **39.943 μs** |  **7.2021** |   **33.3 KB** |
|                    |                    |                    |                |                |                |                |             |               |
|    **IISPreStart** |          **False** |              **5** | **143.702 μs** |  **2.7051 μs** |  **2.5303 μs** | **143.335 μs** | **10.2539** |  **47.97 KB** |
|    **IISPreStart** |          **False** |             **10** | **289.001 μs** |  **5.3822 μs** |  **5.2860 μs** | **287.355 μs** | **20.5078** |  **95.91 KB** |
|    **IISPreStart** |          **False** |             **25** | **728.805 μs** | **12.5603 μs** | **12.8985 μs** | **726.538 μs** | **51.7578** | **239.88 KB** |
|    **IISPreStart** |           **True** |              **5** |  **62.554 μs** |  **0.9244 μs** |  **0.7719 μs** |  **62.470 μs** |  **2.4414** |  **11.69 KB** |
|    **IISPreStart** |           **True** |             **10** | **129.738 μs** |  **2.5527 μs** |  **3.0388 μs** | **130.842 μs** |  **4.8828** |  **23.38 KB** |
|    **IISPreStart** |           **True** |             **25** | **309.804 μs** |  **2.6852 μs** |  **2.0964 μs** | **310.451 μs** | **12.2070** |  **58.48 KB** |

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs [hunk 1]
```diff
@@ -144,7 +219,7 @@ private void RemoveAllCorrelationIdentifierContexts()
             }
         }
 
-        private void SetLogContext(string service, string version, string env, ulong traceId, ulong spanId)
+        private void SetLogContext(ulong traceId, ulong spanId)
```
Comment (zacharycmontoya) [inline src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs:222]: Additional unrelated change I made while I was here: The service, version, and env of the Tracer never change, so I removed those argument from the method.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 2]
```diff
@@ -2150,15 +2161,6 @@ Debug("GenerateVoidILStartupMethod: Linux: Setting the PInvoke native profiler l
     return hr;
   }
 
-  ULONG string_len = 0;
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:2153]: Another unrelated change, `string_contents` was never used anywhere so I deleted. It was probably used for debugging at some point

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 3]
```diff
@@ -712,6 +712,17 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
       Warn("JITCompilationStarted: Call to RunILStartupHook() failed for ", module_id, " ", function_token);
       return S_OK;
     }
+
+    if (is_desktop_iis) {
+      hr = AddIISPreStartInitFlags(module_id,
```
Comment (tonyredondo) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:717]: 😅 another method to port to ReJIT when I start to work in NGEN images support... hehe

**Review by tonyredondo — APPROVED** [review]
LGTM, glad to see the AppDomain.SetData outperforming...

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 4]
```diff
@@ -2417,6 +2419,194 @@ Debug("GenerateVoidILStartupMethod: Linux: Setting the PInvoke native profiler l
...
+  // create the proper string for CreateInstance to successfully call
+#ifdef _WIN32
+  LPCWSTR pre_init_start_str = L"Datadog_IISPreInitStart";
+  auto pre_init_start_str_size = wcslen(pre_init_start_str);
+#else
+  char16_t pre_init_start_str[] =
+      u"Datadog_IISPreInitStart";
+  auto pre_init_start_str_size =
+      std::char_traits<char16_t>::length(pre_init_start_str);
+#endif
```
Comment (kevingosse) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:2529]: At some point we should think of introducing a macro for those

File: src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs [hunk 5]
```diff
@@ -209,5 +284,38 @@ private void SetSerilogCompatibleLogContext(string service, string version, stri
...
+                object state = AppDomain.CurrentDomain.GetData(NamedSlotName);
+                _executingIISPreStartInit = state is bool boolState && boolState;
+            }
+            else
+            {
+                var stackTrace = new StackTrace(false);
+                var initialStackFrame = stackTrace.GetFrame(stackTrace.FrameCount - 1);
+                var initialMethod = initialStackFrame.GetMethod();
+
+                _executingIISPreStartInit = initialMethod.DeclaringType.FullName.Equals("System.Web.Hosting.HostingEnvironment", StringComparison.OrdinalIgnoreCase)
```
Comment (kevingosse) [inline src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs:306]: After the value transitions back to false, can't you unsubscribe the `RefreshIISPreAppState` handler?

File: src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs [hunk 6]
```diff
@@ -32,6 +41,41 @@ internal class LibLogScopeEventSubscriber : IDisposable
...
+            // If the property is not set, we must rely on a different method of determining the state
+            object state = AppDomain.CurrentDomain.GetData(NamedSlotName);
+            if (state is bool boolState)
+            {
+                _performAppDomainFlagChecks = true;
+                _executingIISPreStartInit = boolState;
+            }
+            else
+            {
+                _performAppDomainFlagChecks = false;
```
Comment (kevingosse) [inline src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs:57]: As this can have a performance impact, I believe we should add a warning log when `_performAppDomainFlagChecks` is false

**Comment by kevingosse** [comment]
LGTM, I'd just like to add one log

File: src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs (hunk 5 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs:306]: Yes you're right. I'll add that

File: src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs (hunk 6 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs:57]: Sounds good, will add

File: src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs (hunk 5 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs:306]: @kevingosse do you know if there are any caveats of removing the event handler while the same event handler is still running?

File: build/docker/IIS/build.ps1 [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+# Build Datadog MSI's
+nuget restore $trace_solution
+msbuild $trace_proj /t:BuildCsharp /p:Configuration=Release
+
+msbuild $trace_proj /t:BuildCpp /p:"Configuration=Release;Platform=x64"
+msbuild $trace_proj /t:BuildCpp /p:"Configuration=Release;Platform=x86"
+
+msbuild $trace_proj /t:msi /p:"Configuration=Release;Platform=x64"
+msbuild $trace_proj /t:msi /p:"Configuration=Release;Platform=x86"
```
Comment (lucaspimentel) [inline build/docker/IIS/build.ps1:38]: I think these can all be done in a single line:
```suggestion
msbuild $trace_proj /t:msi /p:"Configuration=Release;Platform=All"
```

File: src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs (hunk 6 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs:56]: @kevingosse I made the change in c5c9c63da39cec5fd1863487a0d79c43e734d46b . Is that good for you?

File: build/docker/IIS/build.ps1 (hunk 7 above)
Comment (zacharycmontoya) [reply build/docker/IIS/build.ps1]: Yep you're right, I've applied this in 5042a44b9088512c61fb99bdd20b4fefdefcd5be

File: src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs (hunk 5 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace/Logging/LibLogScopeEventSubscriber.cs:315]: Done in aacd0a06d838ffbde0cac80270ca73335b6b0f0f

**Review by lucaspimentel — APPROVED** [review]
